### ISSUE TITLE
v15

### ISSUE THREAD

**Issue by SubJunk** [UniversalMediaServer/UniversalMediaServer#5204]
I'm thinking of a group of UI updates for v15. The list can be added to:

- [x] Make the web settings top-right dropdown menu into a left-side menu on desktop instead, to save an extra click
- [x] Move the user avatar on the web player and settings to the top-right where the dropdown menu was
- [x] Make the user actions like login and logout happen from the avatar menu
- [x] Add "switch user" functionality
- [x] Make it clearer when you are using the default/anonymous user
- [x] In the web player, display dynamic TV season information when viewing a season (like a season cover image if TMDB has it, etc., we already have that in the db and API)
- [x] In the web player, add a file media info (location, types, streams, subs, ...)
- [x] Use the custom name of the server on the web player and settings (from https://www.universalmediaserver.com/forum/viewtopic.php?t=16328)
- [x] Make the size of the left menu in the web player customizable by ~~dragging the dividing line~~ (impossible in Mantine, fall back to Customize view), should persist in the browser (from https://www.universalmediaserver.com/forum/viewtopic.php?t=16328)
- [x] Make Mantine scrollbars wider (Other scrollbars (like browser ones) are not under React control) (from https://www.universalmediaserver.com/forum/viewtopic.php?t=16328)
- [x] Add a buffer to the bottom of the left menu in the web player to prevent the scrollbar sitting on top of the last link (from https://www.universalmediaserver.com/forum/viewtopic.php?t=16328)
- [ ] Link to the support site from the web interface
- [x] Check changes on mobile web browser
- [x] Make Shared Content area information more easily seen, e.g. https://github.com/UniversalMediaServer/UniversalMediaServer/issues/4813 and have less padding, I currently have to scroll even though I don't have much shared
- [x] Remove Heapmem use
- [ ] Put series logo and wallpaper on the season page, since it looks empty with just the poster and tagline

**Comment by Mik-S-UMS** [comment]
Someone just asked if it was possible to do some customisation of the web interface [here](https://www.universalmediaserver.com/forum/viewtopic.php?t=16328) and gave some suggestions on improvements.

**Comment by SubJunk** [comment]
Those are all good suggestions, I have added them to the list

**Comment by SubJunk** [comment]
I have a branch in progress for this stuff, so far it is looking like this:

<img width="1411" alt="Image" src="https://github.com/user-attachments/assets/11c23a7b-60c4-4a9f-ae5c-6574a9c33aab" />

<img width="1410" alt="Image" src="https://github.com/user-attachments/assets/c05e77e5-b01b-4f35-bac9-43903bdd4287" />

So there is this concept of two "modes" - Play mode and Manage mode. I think this is more simple for users because then they can match what they want to do with the area, do they want to play media, or manage UMS. 

I'm interested in feedback

**Comment by Mik-S-UMS** [comment]
I tried building with that script but could not remember what I had to change when I did it for V14 beta. 
I thought it was just replacing `main` with `v15` but it is still building 14.10.0-SNAPSHOT
What do I need to do to change branches?

**Comment by SubJunk** [comment]
The version will still say "14.10.0-SNAPSHOT" since I didn't change that yet

**Comment by Mik-S-UMS** [comment]
That was confusing me but I still not sure if it is working right as I cannot see those changes and still looks like 10.14.0 with the avatar on the left.

Am I correct that the code update section in the batch file needs to be `CALL java -jar ../jgit-cli.jar merge origin/v15 -s theirs` to build this branch?

**Comment by SubJunk** [comment]
Hmm yeah that looks like the right command. You might have to also remove the old code too, I don't know if that will switch branches

**Comment by Mik-S-UMS** [comment]
Yeah I tried that. If I didn't it would give build errors so I extracted the build script into a new folder just for v15 but still no luck.

**Comment by SubJunk** [comment]
Can you upload the build script you use? I have one but not sure how similar it is

**Comment by Mik-S-UMS** [comment]
Its the one you sent me in PM on the forum, "UMSBuilder-1.10.0.zip" 
The only change was that line for v15

**Comment by SubJunk** [comment]
Ah I see, I just tested it and the way it is written, it skips the branch part the first time it runs. So you can run it one time, it will compile `main`, then run it again and it should compile v15. I also just pushed a change to set the version in that branch, so it should show up as `15.0.0-ALPHA` now

**Comment by Mik-S-UMS** [comment]
It gets further on the second run and using the right branch but comes up with the error I had previously if the code folder existed.
```
[INFO] --- frontend:1.15.1:yarn (yarn build) @ ums ---
[INFO] yarn not inheriting proxy config from Maven
[INFO] Running 'yarn build' in C:\Users\user\Downloads\UMSBuilder-1.10.0\UMSBuilder-1.10.0\Code\react-client
[INFO] src/components/Home/Home.tsx(17,15): error TS6133: 'Button' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(17,39): error TS6133: 'NavLink' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,42): error TS6133: 'IconHome' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,52): error TS6133: 'IconInfoCircle' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,68): error TS6133: 'IconPlayerPlay' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,84): error TS6133: 'IconSettings' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,98): error TS6133: 'IconShare' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,109): error TS6133: 'IconTool' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,119): error TS6133: 'IconUser' is declared but its value is never read.
[INFO] src/components/Home/Home.tsx(22,129): error TS6133: 'IconUsers' is declared but its value is never read.
[INFO] src/components/Navbar/Navbar.tsx(18,36): error TS6133: 'IconPlayerPlay' is declared but its value is never read.
[INFO] src/UmsApp.tsx(19,10): error TS6133: 'useContext' is declared but its value is never read.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  29.867 s
[INFO] Finished at: 2025-02-16T15:24:36Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.15.1:yarn (yarn build) on project ums: Failed to run task: 'yarn build' failed. org.apache.commons.exec.ExecuteException: Process exited with an error: 2 (Exit value: 2) -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
A duplicate file name exists, or the file
cannot be found.
```
I ran with the -X switch on for a detailed log.

[output.txt](https://github.com/user-attachments/files/18815284/output.txt)

**Comment by SubJunk** [comment]
Ah that makes sense, I have only run it in development mode where it wouldn’t fail for those things. I’ll let you know when I have fixed them

**Comment by SubJunk** [comment]
Does it work now?

**Comment by Mik-S-UMS** [comment]
That is working now and have it installed.

I like that it is easier to switch both modes with a click now and easier access to the settings in the side bar.  

These are some suggestions I have been thinking about.

I use FrankerFaceZ on twitch and that has a good way of showing what options you have changed by highlighting them and show a "x" to restore them to default. Maybe something similar could be use in UMS settings? 
I know that the default settings are sent to the WebGUI anyway so this this should be fairly easy to do.
![Image](https://github.com/user-attachments/assets/e85eb708-1eda-456d-839b-4d6fd082e4b4)

The shared content could be simplified a bit.
At the minute the shared folders path is squashed and there are 2 reset buttons. 
Clicking on the text box does nothing as you cannot type here.
![Image](https://github.com/user-attachments/assets/1282a7ee-660e-4950-b41b-97fddf57e196)
I have been looking as what elements are available to Mantine and think for the folder selection it could be something like [this](https://mantine.dev/core/file-input/#left-and-right-sections) 
![Image](https://github.com/user-attachments/assets/c838b74a-d4bf-4f25-bad8-538994d3dca0)
where you click on the box and opens the folder selector then you don't need the folder button. 
Just replace the icon with something more relevant and then add the "x" afterwards to clear the entry.
The same thing can be used in other sections of the settings too.

The folder selector could do with a button to select the current folder displayed instead of needing to navigate to the parent to select it, like this
![Image](https://github.com/user-attachments/assets/4ffb169a-d85e-44b8-a36f-d91f40b352b7)

For virtual folders you can also get rid of the extra "add folder" under the shared folder, just add another blank input box under the last one once that has something in it.

Maybe all folders could be set up as virtual folders instead of having these 2 options, normal folders will become virtual with only one shared folder. I don't know if there is any technical difference with how UMS handles them now and gives the user the ability to rename them so they show a different name on their device.

The path option could be removed from this too as it would make more sense to have it in the main shared content by nesting them, but not sure if it is possible in Mantine. You could also show icons instead of the folder type.

This is a quick mock-up in mspaint on how it could look

![Image](https://github.com/user-attachments/assets/8cc317ab-e332-45a2-8bb1-a78764aa0b59)

![Image](https://github.com/user-attachments/assets/42e3fee4-8df2-47da-b822-68d143b5d0f5)

**Comment by SubJunk** [comment]
These are great suggestions, I'll go through them and try to figure out what's possible etc

**Comment by Mik-S-UMS** [comment]
I think you should also improve the subtitles settings to make it easier to use as this is not very intuitive to figure out the right language codes and order for specific audio languages priorities. It would be better if the user could just select their language from a list without needing to know the language code, and a better way or arranging language pairs.

Maybe you could have a specific place for all the language settings in one place which could just duplicate the existing settings in various sections (Subs, Audio, and UI)

**Comment by SubJunk** [comment]
@Mik-S-UMS yeah that was actually one of the reasons for making web settings, because that kind of thing is really hard to make in Java but easy in JavaScript. I have an old branch where I tried to make dynamic language dropdowns in the Java GUI before giving up :D

**Comment by SubJunk** [comment]
I'm making progress on the user switcher, will have something to push soon

**Comment by SurfaceS** [comment]
About mobile web browser, I think we can add an option to avoid usage of navbar, in favor of the icon folder (like UPnP) to have a better experience.
Also, it can be useful for TV browsers
I'll have a look on this

**Comment by Mik-S-UMS** [comment]
I think it is about time to add a renderer conf editor in the webUI to make it easier for people to create confs for devices that do not have one yet. 
Maybe you could have an option for the user to submit their conf to an online database so it can be used by others.
This could also collect information (opt in) about the device like the UserAgent, UpnpDetails, ModelName, GetProtocolInfo, etc that can be used to improve support for other devices.
This will allow us to support even more devices even if we do not have access to them. 

Ideally most of the setup could be automatic by asking the user to play various test videos on the device and report if they play, then this can be refined by showing more advanced options.
.
.
.
I was recently [asked](https://www.universalmediaserver.com/forum/viewtopic.php?t=16505&sid=cea4c397438358929acaf135c5b72bc1) on what makes the decision on the format to transcode in as UMS was using h265 when transcoding to their TV. Some people will not have a powerful enough computer to do this.
This does make sense if the video is in h265 so it only needs to be remuxed if the audio format is not supported, however h264 would be transcoded to h265 as I understand it.
Maybe the transcoding profiles could be dynamic so the best one is chosen based on the video format and the capabilities of the computer.

**Comment by SubJunk** [comment]
Great progress here thanks! I have been low on time lately but I have some time today to make more progress on the user switcher

**Comment by SubJunk** [comment]
Another simple idea I had is with the taskbar menu, we have:

- Web player
- Settings
- Settings (old)
- Quit

But I think the "web player" is confusing because we have two web players - an anonymous player, and a user-specific player. So I am thinking of other words, and feedback is welcome. 
The anonymous player can be confusing I think because it has no way to log in, so users might miss our user-specific features.

Here's a first attempt, where the top option is a new one:
- Web player
- Web player (anonymous)
- Settings
- Settings (old)
- Quit

**Comment by SurfaceS** [comment]
added this `file info` to the web player :
This should help users/members to investigate things.
![Image](https://github.com/user-attachments/assets/7d81c2ee-2af8-496e-93d9-e8d385809f14)

![Image](https://github.com/user-attachments/assets/751db9b8-33b4-4b12-afc8-080ee6d4296e)

![Image](https://github.com/user-attachments/assets/88eddc62-5892-48a7-81c1-75c7633d9d56)

**Comment by SurfaceS** [comment]
> But I think the "web player" is confusing because we have two web players - an anonymous player, and a user-specific player. So I am thinking of other words, and feedback is welcome. The anonymous player can be confusing I think because it has no way to log in, so users might miss our user-specific features.
> 
> Here's a first attempt, where the top option is a new one:
> 
>     * Web player
> 
>     * Web player (anonymous)
> 
>     * Settings
> 
>     * Settings (old)
> 
>     * Quit

@SubJunk 
What you call `Web player (anonymous)` is not anonymous. It can use auth, etc like `Settings` player do.

It is more `Web player only`, when the `Settings` is `Web player and settings` or `UMS interface`
I'm not a native English speaker, but you know what I mean.

---



@Mik-S-UMS 
I added the file info to the web player (previous post) to help people and helpers know the codec etc...

> I think it is about time to add a renderer conf editor in the webUI to make it easier for people to create confs for devices that do not have one yet.

This is a huge work.
Even the one that was created on the java GUI was never helpful to me.

The problem with the current implementation is that it do not split the browser and the player.
In UPnP it is the Control Points (browser), MediaRenderer (player).

Normally, everything advertised is what you get.
Here, we advertise something adapted to `Control Point`, and adapt it again for the `MediaRenderer`.

I'll have a look on what can be done on this side.

**Comment by SubJunk** [comment]
> What you call Web player (anonymous) is not anonymous. It can use auth, etc like Settings player do.
It is more Web player only, when the Settings is Web player and settings or UMS interface
I'm not a native English speaker, but you know what I mean.

Your English is great. Why do we need both of the web players? I think it is confusing

**Comment by SurfaceS** [comment]
> Why do we need both of the web players? I think it is confusing

For security concerns.

The 9001 server handle every part of the react GUI.
The 9002 server only handle player things. It do not run the settings part on UMS side. Then it cannot be attacked on the settings side.

The 9002 can be shared inside and outside the network without been afraid someone will change settings, restart the computer and so on (accidentally or in a hacking way).

**Comment by Mik-S-UMS** [comment]
@SurfaceS That is really useful and will help a lot.
There is not really a conf editor in the java GUI, it only opens a text editor with for that one specific device using the uuid or IP and requires the user to know how to write a conf and reference the DefaultRenderer.conf for the documentation of all the settings.
The few times I have seen it used it has actually caused more problems than tweaking an existing conf.

I'm thinking of a proper GUI where you can pick what file types and formats the device is able to play and this will populate a table for the supported formats section which can further refined in an advanced mode if needed. 

The render matching info could be taken directly from a detected device on the network where you can just select the model number or some other unique information. This way you do not need to search the logs for User-Agent or other UPnP details.

The user would not need to know what each format tag as this may be listed as many things. For example they may not know that AC3 is Dolby digital, that E-AC3 is Enhanced AC3 or Dolby Digital Plus or DD+ etc, so It would be good to have these all available with the full names and aliases. Mime types could also be added automatically as they should be standard for that file type, but could also be changed in an advanced mode.

An existing conf could be imported if there only needs to be a minor tweak like if a manufacturer drops support for a format (like DTS) in the new models but otherwise the would have the same capabilities as the previous models, then you could just untick that format instead of needing to manually remove from each line.

Also as each option is available the documentation can also be transferred to the tooltips which can also be localised for other languages. 

I know this would take a lot of work so maybe it can be done in small chunks, just a basic selector at first but then you could add some automation and some checking to validate if it is set up correctly by playing test videos which could be downloaded separately.

One major benefit of this would allow a user to submit their conf along with their device details so that better confs can be made for more devices without us needing to to have physical access to them.

**Comment by SurfaceS** [comment]
@Mik-S-UMS 
:thinking::roll_eyes::woozy_face:

1st a cleanup and an interface... :+1:

**Comment by SubJunk** [comment]
> What you call Web player (anonymous) is not anonymous. It can use auth, etc like Settings player do.

@SurfaceS how can users switch user on the port 9002 player? I don't see a way

> The 9002 can be shared inside and outside the network without been afraid someone will change settings, restart the computer and so on (accidentally or in a hacking way).

@SurfaceS I can see the idea and it's good.

I am undecided on the way forward, because I think it is confusing as a user, so if we keep the port 9002 one it needs to be explained to the user better, with more intuitive naming and probably a Knowledge Base article.

Some options I can think of are:

1. Start calling it the "standalone" player:

    * Web player
    * Web player (standalone)
    * Settings
    * Settings (old)
    * Quit

2. Make the standalone player hidden from tray menus by default but people can still access it. So the tray menu is the same as it is currently, but "Web player" links to the port 9001 player:

    * Web player
    * Settings
    * Settings (old)
    * Quit
3. Make the standalone player disabled by default but people can enable it in settings
4. Remove the standalone player, because JWTs are good and we keep our libraries and security practices up to date. I think we have good enough security to trust it. But I won't remove it if you think it is important.
5. Another idea I didn't think of?

**Comment by SurfaceS** [comment]
@Mik-S-UMS 
I created a new issue for this, as it will not be released soon, even if I hard work on it.

@SubJunk 

> how can users switch user on the port 9002 player? I don't see a way

![Image](https://github.com/user-attachments/assets/fd2bc0a3-bb8a-4810-8e19-fd9f34ffcddd)


> Remove the standalone player, because JWTs are good and we keep our libraries and security practices up to date. I think we have good enough security to trust it. But I won't remove it if you think it is important

Even if we do the best for security in the java part, it is better to keep it as another wall.
No code access is better than secured code.

For an internet share, only the 9002 port will be opened on the router, but not the 9001. So no access to settings part.
For a local network, if you create a child account for example, it will have access to the settings interface, so it need to be secured as well.

If I remember well :
- We add the react GUI to replace the old player.

- Then we put setting in it to "replace" java GUI / add a new way to setup UMS.

- As some users was sharing the old player thru internet, we created the standalone player to reflect what was before and for security reason.

Perhaps from the beginning we should have separated the settings part from the player part.
Then we have settings 9001 and player 9002.
But... the player in the settings part is not so bad, as we can check the media directories, change metadata, and so on.

In my head : 9001 is the settings, 9002 is the player.

So I don't think changing the menu will help, and we should leave it like it is :hand_over_mouth:

Maybe the only things that should be done is a better documentation on the standalone player settings.

**Comment by SubJunk** [comment]
@SurfaceS Hmm, maybe we should enable the authentication service on the web player by default.

Then we still have the problem that sometimes a user is on the standalone web player, and sometimes on the web player with settings. Users don't always know about ports, especially if they are only using UMS within their local network, I think they will wonder "why can I sometimes see settings and sometimes not? why can I sometimes login/logout but sometimes not?"

I think the situation works for more advanced users who are forwarding port 9002, but I think we have a lot of users who are not advanced and just use defaults. And I think those users are the ones who need the most help with UX design. Advanced users will hack and find a way to do what they want, and read documentation. 

I have another idea that might help. We could make the standalone player interface the same as the other one, but if it is on port 9002 the new "Manage" button is disabled, and there is a tooltip or modal window to explain why it is disabled for that port.

We could also detect whether the user is accessing it locally or via internet, and if it is local, offer to redirect them to the port 9001 manage area. Or even automate that; if we are only worried about users who access it via the internet, it should be safe to redirect local users.

Mostly I want users by default to be able to switch between web player and web settings using our new mode switcher.

Do you have any thoughts about it @Mik-S-UMS since you talk a lot with our users?

**Comment by Mik-S-UMS** [comment]
TBH I forget about the standalone player on 9002 and I am not sure if anyone uses it even when accessing from outside the network, at least I have not seen anyone mention it.

Maybe only the standalone should be accessible from outside the network so settings can't be changed or activate the restart or shutdown options. There definitely should be some authentication on this otherwise anyone could access your media library if they know your IP. Settings could be locked to only the local network, there should not be any need to change them remotely.

We do need some documentation on how to set up UMS for use outside the local network. I know that some ports need forwarding at least but some users don't want to open up ports and have used other methods like proxies or DNS. These are things I know nothing about so I cannot really help them when their setup is causing problems. 

To me the best solution for this is to use a proper VPN so the remote computer appears as local but most people will use some commercial VPN that won't work for this. So far I have only seen one person using this method as they were watching videos while on holiday in another country. They were the one who reported the marking not working in the web player.

**Comment by SubJunk** [comment]
I'm not sure if this is a bug or because I'm changing versions.

This is what I saw after changing from v14 to v15:
![Image](https://github.com/user-attachments/assets/8e950448-2287-4a08-a259-229f7662da7a)
and if I click "Log in" it selects the "User" part so I have no way to complete the selection. But then shortly after I took this screenshot, it changed to this:
![Image](https://github.com/user-attachments/assets/41850ca8-74c0-4530-9fcd-c143c98a010e)

So maybe this is a problem with session clearing or something, or I'm in a bad state via testing

**Comment by SurfaceS** [comment]
@SubJunk 
> So maybe this is a problem with session clearing or something, or I'm in a bad state via testing

Do you run the latest v15 ?
Last week I merged that
https://github.com/UniversalMediaServer/UniversalMediaServer/commit/96e8fc8f0b100742c3f5beb6d87633703f6a47f0#diff-9d3279eceb0e25fa2442a59c8479742c5df7be62dff438bf7ba71fe3d0443ad6

```
react web : do not put token if expired
added try catch for InvalidTokenError
```
I'll have a look on the code

**Comment by SubJunk** [comment]
> Do you run the latest v15 ?

Yeah this was in 824f19d7525239cde97ed94ab40604f40d2a4068 and it happened again tonight, so it seems it happens each day

**Comment by SurfaceS** [comment]
@SubJunk I found it.
fixed in #5439 + other fix

**Comment by SubJunk** [comment]
Cool I will test it, nice work

**Comment by SubJunk** [comment]
The fix works

**Comment by SubJunk** [comment]
I think the way we are using `Card` with `onclick` for video thumbnails is not accessible or keyboard navigable, we should fix that by using an anchor or button wrapper

**Comment by SubJunk** [comment]
Also I am getting FFmpeg errors in the logs while browsing YouTube channels, and the videos do not play. I haven't compared it to `main` yet

**Comment by SubJunk** [comment]
This was done and released
